recall/ipld/src/hamt/core.rs

```diff
@@ -219,6 +231,42 @@ where
         }
     }
 
+    fn inner_for_each_ranged<F>(
+        &self,
+        starting_key: Option<&hamt::BytesKey>,
+        max: Option<usize>,
+        mut f: F,
+    ) -> Result<(usize, Option<hamt::BytesKey>), Error>
+    where
+        F: FnMut(&hamt::BytesKey, &V) -> anyhow::Result<bool>,
+    {
+        let mut iter = match starting_key {
+            Some(key) => self.hamt.iter_from(key)?,
+            None => self.hamt.iter(),
+        }.fuse();
+
+        let mut traversed = 0usize;
+        let limit = max.unwrap_or(usize::MAX);
+        loop {
+            if traversed >= limit {
+                break
+            }
+
+            match iter.next() {
+                Some(res) => {
+                    let (k, v) = res?;
+                    if !(f)(k, v)? {
```

Comment:
if the flag is `false`, we go to the next element and don't increase `traversed`. Take a look at the [original implementation](https://github.com/filecoin-project/ref-fvm/blob/master/ipld/hamt/src/hamt.rs#L424-L449), see how the use of the flag allows us to skip elements without increasing `traversed`